### ISSUE TITLE
Rule Change: no-implicit-coercion should suggest parseInt/parseFloat instead of Number

### ISSUE THREAD

**Issue by stianjensen** [eslint/eslint#15734]
### What rule do you want to change?

no-implicit-coercion

### What change to do you want to make?

Implement suggestions

### How do you think the change should be implemented?

A new option

### Example code

```js
const myNumber = "10";
const sum = +myNumber + 50;
```

### What does the rule currently do for this code?

The rule no-implicit-coercion will present an error with the following suggestion:
> use `Number(myNumber)` instead

### What will the rule do after it's changed?

The rule no-implicit-coercion will present an error with the following suggestion:
> use `parseFloat(myNumber)` instead

### Participation

- [X] I am willing to submit a pull request to implement this change.

### Additional comments

At least when working with TypeScript, parseInt and parseFloat are typed to only accept strings:
https://github.com/microsoft/TypeScript/issues/4002
While Number is typed to accept any.

So when working with typescript, unary + and Number are equally unsafe, while parseInt and parseFloat actually catch errors when you accidentally provide a string (say a whole object instead, where you should have provided a specific key from that object).

My suggestion is to either update the recommended fix, or add an option to override it. Probably also to allow you to either disable the autofix, or have it fix to your wanted recommendation. I assume that just changing the default fix from Number to parseFloat is probably not feasible, wrt back-compat.

**Comment by mdjermanovic** [comment]
Hi @stianjensen, thanks for the issue!

This is a stylistic rule that disallows shorthand type conversions like `!!foo` or `+bar` in favor of their _equivalent_ explicit forms like `Boolean(foo)` or `Number(bar)`.

`parseFloat(myNumber)` is not the same as `+myNumber`. For example, it can parse only a part of  the string:

```js
const myNumber = "1.2bar";

+myNumber; // NaN
Number(myNumber); // NaN
parseFloat(myNumber); // 1.2
```

Therefore, I think that suggesting `parseFloat(myNumber)`, which is significantly different from `+myNumber` / `Number(myNumber)`, should be out of scope for this particular rule. The use case you described looks to me like a case for another rule that would disallow both `+myNumber` and `Number(myNumber)` in favor of `parseFloat(myNumber)`.

**Comment by stianjensen** [comment]
Hi, @mdjermanovic and thanks for the thorough answer!

The difference in behavior makes sense that it is more of a transformation than may be safe to convert automatically.

I'll look at creating a plugin for providing a rule like this instead, then!
